### ISSUE TITLE
SetMixerFPUMode calls fesetround(FE_TOWARDZERO) without restoring in RestoreFPUMode

### ISSUE THREAD

**Issue by flibitijibibo** [kcat/openal-soft#148]
Title says most of it, but here are some code snippets for context:

https://github.com/kcat/openal-soft/commit/be4a4d74686fffb0a7e38cb4b1cad9653d2ba00f

https://github.com/kcat/openal-soft/blob/be4a4d74686fffb0a7e38cb4b1cad9653d2ba00f/Alc/helpers.c#L211

Simply put, the calls to store/restore mode were replaced, but the actual call that changes the mode stayed in the code and hasn't been touched since. Miraculously this only appears to have mattered since ~1.17, and only on Windows!

I fixed it locally with this patch, but we should probably do proper store/restore to fix this:

```
diff --git a/Alc/helpers.c b/Alc/helpers.c
index 8e6306ae..e7a78999 100644
--- a/Alc/helpers.c
+++ b/Alc/helpers.c
@@ -346,6 +346,10 @@ void RestoreFPUMode(const FPUCtl *ctl)
         __asm__ __volatile__("ldmxcsr %0" : : "m" (*&ctl->sse_state));
 #endif
 
+#ifdef FE_TOWARDZERO
+    fesetround(FE_TONEAREST);
+#endif
+
 #elif defined(HAVE___CONTROL87_2)
 
     int mode;
```

**Comment by kcat** [comment]
Odd. I'd have expected ```fegetenv``` and ```fesetenv``` to save and restore the floating-point rounding mode.

**Comment by flibitijibibo** [comment]
Yeah, every page about it says the same thing, which made me skeptical at first. I wonder if it's returning an error for some obscure reason and we don't know it...? I'd be surprised if this was a runtime bug.

If it's any help, this is with the official releases as well as my own builds (Fedora 26 MinGW and CentOS 7 MinGW).

**Comment by flibitijibibo** [comment]
Just took a look at Microsoft's page, maybe this is it?

https://docs.microsoft.com/en-sg/cpp/c-runtime-library/reference/fesetenv1

    To use this function, you must turn off floating-point optimizations that could prevent access by using the #pragma fenv_access(on) directive prior to the call. For more information, see fenv_access.

**Comment by kcat** [comment]
Unlikely, since that seems to be related to compiler optimizations, which would affect MSVC rather than MinGW. I tested it myself, and get appropriate results on Linux (the rounding mode is properly reset after ```fesetenv```), while the Windows build built with MinGW keeps the modified value.

Out of curiosity, have you tried testing on real Windows? I only have Wine for testing, and it's possible Wine's implementation of ```fegetenv```/```fesetenv``` are buggy and not properly handling the rounding mode.

**Comment by flibitijibibo** [comment]
Can confirm this happens on Windows 7 and 10 at minimum - it happens in Wine for me too, so it appears Wine is accurate here! I wonder what it's doing to make Linux reproduce the issue.

**Comment by kcat** [comment]
Hmm, one other possibility is that it's a problem in MinGW. Code built with MinGW does not rely on MS's newer C runtimes (msvcr100, 110, etc), and msvcrt isn't wholly reliable for the newer C99 and C11 functions, so it provides them via an automatically linked static library. Assuming I'm reading this correctly, the output of ```nm``` seems to suggest the function is built into the DLL rather than being imported from another. If that's the case then it won't matter if it's run on Wine or real Windows, the DLL is using an internal (buggy) version of the function provided by MinGW.

I'll need to make a workaround for now regardless, but it seems like something MinGW should fix.

**Comment by kcat** [comment]
A workaround was added with commit bc386af5c5607f0d7a7ed911d5975fc9758c7110. On Windows, it explicitly saves and restores the rounding mode.

**Comment by flibitijibibo** [comment]
Good enough for me! Looked up the MinGW source as well as the Wine source and found this in case anyone wants to fix this for reals:

https://github.com/mirror/mingw-w64/blob/1e81200e88f19e440bbe646bfb410159ff8b0d51/mingw-w64-crt/misc/fesetenv.c

https://github.com/wine-mirror/wine/blob/302153117e20b62c9170aed62aa33e83cacfaf59/dlls/msvcrt/math.c#L1365

**Comment by matpow2** [comment]
This is a bit scary. Was this also an issue in the OpenAL-Soft Win32 32bit binaries distributed here? http://kcat.strangesoft.net/openal.html

**Comment by kcat** [comment]
Most likely yes, since they're built with MinGW-w64. I'm honestly quite surprised I never heard of or saw any issue related to this prior to this report. I guess this means I'll be needing to make a 1.18.2 release quite soon with this fix, probably branched off from the 1.18.1 release so as to not include too many of the more drastic changes for a bug-fix release.

**Comment by Marisa-Chan** [comment]
Just want notice about fesetenv was finally fixed in mingw-w64 in 2019 : https://sourceforge.net/p/mingw-w64/mingw-w64/ci/d5274305abebaad50564a0778f01d2d6556ad2ec/
